### ISSUE TITLE
Synchronizing colors:

### ISSUE THREAD

**Issue by jordwalke** [onivim/oni#33]
I [commented about synchronizing keyboard shortcuts](https://github.com/extr0py/oni/issues/20) and the various options for doing so.

TL;DR:
> - No one wants to write VimScript (how can I blame them).
> - Everyone knows JS/JSON so it is preferable.
> - But we don't want a ton of startup time overhead to wait for a JS engine to warm up or wait for latency to cross the language bridge several times (if we can avoid it).
> - We also don't want half our editor config in VimScript and half in JS/JSON long term so if there are two ways to do something it would be nice if there was a migration path to have everything in one language, without paying any performance costs (a suggested approach is ahead of time compiling JS/JSON to VimScript - which also lets you use that same config in any build of Vim).

Now, I'd like to make an Issue to track the same kind of configuration experience for color schemes and fonts including colors of objects that don't have a corresponding element inside of stock Vim. (popup menus etc). It would be nice to configure everything in one place, just like keyboard mappings. It would also be great if every color scheme could work out of the box with UI extensions with a nice default color that synced to the color scheme bgcolor etc - but then *also* allow colorschemes to add customizations for these UI extensions. It would even be preferable if color schemes could be *defined* in JS, and then executed ahead of time to spit out regular vimscript config. Imagine if all the color blending logic was done in JS, and the output is raw vim highlight commands (this is much like our shift to css in JS).

**Comment by extr0py** [comment]
Great feedback. I've been wondering how to handle color schemes. One option was just to stick with the default Vim colorschemes. 

I was taking a look at some other editors to see how they define colorschemes:
1) VSCode - uses TextMate themes
2) Atom - uses less files (makes sense based on the way it renders)

I thought at first about just using the vim color themes, but we run into the problem where the UI extension don't respect the themes at all. One option I looked briefly at was reading the set of highlights and applying them to the UI - trying to pick good defaults for the built-in UI elements like the menu, etc. 

I like the idea of having the colorscheme configuration be in JS too though, to be consistent with the other configuration. The hardest part with that is to bridge the gap so that users can bring in the themes they are comfortable with. 

So from that that, a `colors.js` file might look like this:
```
const config = {
"base": "onedark.vim",
"ui.bgColor": "black",
"ui.autoComplete.border": "gray",
"ui.autoComplete.background": "rgba(200, 200, 200, 0.5)",
..etc...
} 

export default config
```
And once we've parsed this file, we'd convert it the vim-syntax related items to highlight instructions, and preserve the rest to be used in the UI. Is that sort of what you were thinking?

**Comment by jordwalke** [comment]
> highlights and applying them to the UI - trying to pick good defaults for the built-in UI elements like the menu, etc.

I do think that is a good approach. Use the popup menu for the "command p" menu colors for example.

> The hardest part with that is to bridge the gap so that users can bring in the themes they are comfortable with.

You could use any existing Vim color scheme, but then just provide a JSON/JS API for *creating* those (again, could be converted to plain VimScript ahead of time to not block startup time). I think you could take the top 5 Vim colorschemes and port them from VimScript to JS(ON) by hand just so that they serve as an example. Once you give people a JS API to do things, the community tends to come and fill in all the missing pieces for you.  I could see them creating all kinds of tools to port over VimScript syntax themes to the JSON format (which is then transformed back into VimScript behind the scenes). I think Atom only ships with 5 stock colorschemes, but look what people have done! I would probably help you port a few of my favorite ones.

> And once we've parsed this file, we'd convert it the vim-syntax related items to highlight instructions, and preserve the rest to be used in the UI. Is that sort of what you were thinking?

Yes, except I might suggest that you try to define the entire color scheme in JS instead of supporting the "extension" of the base `.vim` colorscheme. It just keeps it simple and makes your job much easier. Because people like JS, you don't need to worry - people will port everything to JS. Everything that can be ported to JS *will* be ported to JS! :D

And I do believe that converting these *super* high priority configuration should be cached so that the second time you start up the editor, you do not need to recompute them (unless the file changed.) Baking that behavior in early is very wise because it ensures people don't add features that degrade startup time.

**Comment by teto** [comment]
A PR that can be of interest regarding this issue https://github.com/neovim/neovim/pull/7082

**Comment by bryphe** [comment]
@teto mentioned in chat that this commit came in: neovim/neovim@db99982, this could be helpful for synchronization too. Hooking the `ColorScheme` autocmd and then querying for the highlight groups would allow Oni to tap into some colors besides just the foreground/background color. 

An example would be using the corresponding highlight color in auto completion (instead of just that hardcoded green color, we could color function completions with the function highlight group color, etc).

**Comment by bryphe** [comment]
Finally starting to make progress on some of these year-old issues! 👍 Thanks for all the thinking around this!

An initial json theming implementation was included with #950 (in PR #975), and is customizable both via `json` themes associated with a vim theme... as well as configurable by `colors.` directly in configuration:


Here's the wiki additions:
- [Theming](https://github.com/onivim/oni/wiki/Theming)
- [Customizing colors via config](https://github.com/onivim/oni/wiki/Configuration#colors)

These updates will be included in the next release (0.2.18).

And then there are a few items still being tracked for a complete story:
- #412 - CSS Theming
- #1051 - Colorization based on TextMate scopes

**Comment by bryphe** [comment]
I'll close this out, and track the remaining work in those other items, but of course any feedback is welcome!
